contracts/FeralfileEnglishAuction.sol

```diff
@@ -0,0 +1,378 @@
+        uint256 extendDuration;
+        uint256 extendThreshold;
+        uint256 minIncreaseFactor;
+        uint256 minIncreaseAmount;
+        uint256 minPrice;
+        bool isSettled;
+    }
+
+    struct Bid {
+        address bidder;
+        uint256 amount;
+        bool fromFeralFile;
+    }
+
+    struct LatestAuctionStatus {
+        Bid latestBid;
+        uint256 endAt;
+        bool isSettled;
+    }
+
+    mapping(bytes32 => EnglishAuction) public auctions;
+    mapping(bytes32 => Bid) public auctionLatestBid;
+
+    constructor(address signer_) ECDSASigner(signer_) {}
+
+    function getAuctionID(
+        address tokenContractAddr_,
+        uint256[] memory tokenIDs_
+    ) public view returns (bytes32) {
+        // order the tokenIDs based on certain rule
+        for (uint i = 0; i < tokenIDs_.length - 1; i++) {
+            uint minIndex = i;
+            for (uint j = i + 1; j < tokenIDs_.length; j++) {
+                if (tokenIDs_[j] < tokenIDs_[minIndex]) {
+                    minIndex = j;
+                }
+            }
+            (tokenIDs_[i], tokenIDs_[minIndex]) = (
+                tokenIDs_[minIndex],
+                tokenIDs_[i]
+            );
+        }
+
+        // hash the (contractAddr + ordered tokenIDs)
+        bytes32 message = keccak256(
+            abi.encode(block.chainid, tokenContractAddr_, tokenIDs_)
+        );
+
+        // return hash as identifier
+        return message;
+    }
+
+    function auctionOngoing(bytes32 aucId_) public view returns (bool) {
+        // check the auction status and return
+        require(
+            auctions[aucId_].id != 0,
+            "FeralfileEnglishAuction: auction not found"
+        );
+
+        return
+            auctions[aucId_].startAt <= block.timestamp &&
+            auctions[aucId_].endAt >= block.timestamp;
+    }
+
+    function setEnglishAuctions(
+        EnglishAuction[] calldata auctions_
+    ) external onlyOwner {
+        // for-loop to set the auction
+        // able to reset an auction if there's no winner and highest bid = 0
+        for (uint256 i = 0; i < auctions_.length; i++) {
+            EnglishAuction memory auc_ = auctions[auctions_[i].id];
+
+            require(
+                auc_.id == 0,
+                "FeralfileEnglishAuction: auction already exist"
+            );
+
+            require(
+                auctions_[i].startAt >= block.timestamp,
+                "FeralfileEnglishAuction: auction start time should be in the future"
+            );
+
+            require(
+                auctions_[i].endAt > auctions_[i].startAt,
+                "FeralfileEnglishAuction: auction end time should be after start time"
+            );
+
+            require(
+                auctions_[i].extendDuration >= 0,
+                "FeralfileEnglishAuction: auction extend duration should be greater than or equal 0"
+            );
+
+            require(
+                auctions_[i].extendThreshold >= 0,
```

Comment:
The same [here](https://github.com/bitmark-inc/feralfile-exhibition-smart-contract/pull/28/files#r1375774811)